crates/subspace-farmer/src/piece_cache.rs

```diff
@@ -0,0 +1,666 @@
+    worker_sender: mpsc::Sender<WorkerCommand>,
+}
+
+impl PieceCache {
+    /// Create new piece cache instance and corresponding worker.
+    ///
+    /// NOTE: Returned future is async, but does blocking operations and should be running in
+    /// dedicated thread.
+    pub fn new<NC>(node_client: NC, peer_id: PeerId) -> (Self, CacheWorker<NC>)
+    where
+        NC: NodeClient,
+    {
+        let caches = Arc::default();
+        let (worker_sender, worker_receiver) = mpsc::channel(WORKER_CHANNEL_CAPACITY);
+
+        let instance = Self {
+            caches: Arc::clone(&caches),
+            worker_sender,
+        };
+        let worker = CacheWorker {
+            peer_id,
+            node_client,
+            caches,
+            worker_receiver: Some(worker_receiver),
+        };
+
+        (instance, worker)
+    }
+
+    /// Check whether piece exists in cache
+    pub fn contains_piece(&self, key: RecordKey) -> bool {
+        // It is okay to take read lock here, writes locks are very infrequent and very short
+        for cache in self.caches.read().iter() {
+            if cache.stored_pieces.contains_key(&key) {
+                return true;
+            };
+        }
+
+        false
+    }
+
+    /// Get piece from cache
+    pub async fn get_piece(&self, key: RecordKey) -> Option<Piece> {
+        let caches = Arc::clone(&self.caches);
+
+        // TODO: Join on drop for this async task
+        let maybe_piece_fut = tokio::task::spawn_blocking({
+            let key = key.clone();
+            let worker_sender = self.worker_sender.clone();
+
+            move || {
+                for (disk_farm_index, cache) in caches.read().iter().enumerate() {
+                    let Some(&offset) = cache.stored_pieces.get(&key) else {
+                        continue;
+                    };
+                    match cache.backend.read_piece(offset) {
+                        Ok(maybe_piece) => {
+                            return maybe_piece;
+                        }
+                        Err(error) => {
+                            error!(
+                                %error,
+                                %disk_farm_index,
+                                ?key,
+                                %offset,
+                                "Error while reading piece from cache, might be a disk corruption"
+                            );
+
+                            if let Err(error) =
+                                worker_sender.blocking_send(WorkerCommand::ForgetKey { key })
+                            {
+                                trace!(%error, "Failed to send ForgetKey command to worker");
+                            }
+
+                            return None;
+                        }
+                    }
+                }
+
+                None
+            }
+        });
+
+        match AsyncJoinOnDrop::new(maybe_piece_fut).await {
```

Comment:
What would be the difference if we just wait for the future?

Reply:
Background thread created by `tokio::task::spawn_blocking` will exist after the future we were awaiting for is just dropped arbitrarily, which in turn would mean `caches` is also kept in memory, which may contain cache backends, which contain open file and memory-mapped descriptors, both of which mean file is occupied. So you might think everything is dropped and files are free, while they are in fact not free and are actively being read in that thread.